django/db/backends/base/base.py

```diff
@@ -270,6 +270,8 @@ def check_settings(self):
     @async_unsafe
     def ensure_connection(self):
         """Guarantee that a connection to the database is established."""
+        if self.in_atomic_block and self.closed_in_transaction:
+            raise ProgrammingError("Cannot open a new connection in an atomic block.")
```

Comment:
This is untested.

Reply:
As far as I'm aware, we don't need this because these three conditions:
- `in_atomic_block is True`
- `closed_in_transaction is True`
- `connection is None`

cannot be fulfilled at the same time (at least in normal flow).

Reply:
I couldn't figure out how to make the situation happen naturally 👍 so makes sense to me

Reply:
This resulted (at least in earlier versions) in a testsuite failure where the atomic behavior was tested. Namely psycopg by default would raise an error if you did something on a closed connection. Now that `close` sets it to `None` it transparently opened a new one :/ This was visible if the Django testsuite was ran with a settings_postgresql that had pooling enabled.

The test in question was `test_atomic_prevents_queries_in_broken_transaction_after_client_close`.

Reply:
Thanks :+1: Now I'm smarter. Let me move this to the `if connection is None` branch.

Reply:
It might make sense to copy the test in question into the postgresql pool specific tests as well and ensure that it works properly under a pool (even when the test settings themself do not run with a pool?)